File: src/python/api/editor.py
from flask import Blueprint, request, jsonify, Response

from ..auth import auth
from src.python.engine.db import Db

api_editor = Blueprint('editor', __name__, url_prefix="/api/editor")


@api_editor.route("/", methods=["POST", "PUT", "DELETE"])
@auth.login_required
def r_editor():
    r = request.json

    if request.method == "POST":
        return jsonify(Db.parse_cond(r.get("cond"), r["offset"], r["limit"]))

    elif request.method == "PUT":
        if r.get("create"):
            if isinstance(r["create"], list):
                c_ids = Db.insert_many(r["create"])
                return jsonify({
                    "ids": c_ids
                })
            else:
                c_id = Db.insert_many([r["create"]])[0]
                return jsonify({
                    "id": c_id
                })

        if r.get("update"):
            if r.get("ids"):
                Db.update_many(r["ids"], r["update"])
            else:
                Db.update_many([r["id"]], r["update"])

        return jsonify({"error": None})

    elif request.method == "DELETE":
        if r.get("ids"):
            Db.delete_many(r["ids"])
        else:
            Db.delete_many([r["id"]])

        return jsonify({"error": None})

    return Response(status=404)


@api_editor.route("/editTags", methods=["PUT", "DELETE"])
@auth.login_required
def r_editor_edit_tags():
    d = request.json
    if request.method == "DELETE":
        Db.remove_tags(d["ids"], d["tags"])
    else:
        Db.add_tags(d["ids"], d["tags"])

    return jsonify({"error": None})

File: src/node/api/editor.ts
import { Router } from "express";
import needUserId from "../middleware/needUserId";
import asyncHandler from "express-async-handler";
import Database from "../engine/db";
import auth from "../auth/token";

const router = Router();

router.use(auth.optional);
router.use(needUserId());

router.post("/", asyncHandler(async (req, res) => {
    const {cond, offset, limit} = req.body;
    const db = new Database();
    return res.json(await db.parseCond(res.locals.userId, cond, {offset, limit}));
}));

router.put("/", asyncHandler(async (req, res) => {
    const {id, ids, create, update} = req.body;
    const db = new Database();
    if (Array.isArray(create)) {
        const ids = await db.insertMany(res.locals.userId, create);
        return res.json({ids});
    } else if (create) {
        const ids = await db.insertMany(res.locals.userId, [create]);
        return res.json({id: ids[0]});
    } else if (ids) {
        return res.json(await db.updateMany(ids, update));
    } else {
        return res.json(await db.updateMany([id], update));
    }
}));

router.delete("/", asyncHandler(async (req, res) => {
    const {id, ids} = req.body;
    const db = new Database();
    if (ids) {
        return res.json(await db.deleteMany(ids));
    } else {
        return res.json(await db.deleteMany([id]));
    }
}))

router.put("/editTags", asyncHandler(async (req, res) => {
    const {ids, tags} = req.body;
    const db = new Database();
    return res.json(await db.addTags(ids, tags));
}));

router.delete("/editTags", asyncHandler(async (req, res) => {
    const {ids, tags} = req.body;
    const db = new Database();
    return res.json(await db.removeTags(ids, tags));
}))

export default router;
